modules/core/src/convert.dispatch.cpp

```diff
@@ -278,8 +278,16 @@ void Mat::convertTo(OutputArray dst, int type_, double alpha, double beta) const
     int dtype = CV_MAKETYPE(ddepth, cn);
 
     Mat src = *this;
-    dst.create(dims, size, dtype);
-    Mat dstMat = dst.getMat();
+    Mat dstMat;
+    if (this->data == dst.getMat().data && this->elemSize() == CV_ELEM_SIZE(dtype)) {
```

Comment:
There is no check for size/dims match. For example, ROI will have the same data ptr:
```cpp
Mat m (10, 10, CV_8U);
Mat roi = m.rowRange(0, 2);
m.convertTo(roi, CV_8S);
```

Reply:
I had through **!dst.getMat().isSubmatrix()** to determine this case.

Reply:
What about the following case? `.data` matched but not the shapes.
```cpp
Mat m (10, 10, CV_8U);
Mat m2(5, 10, CV_8U, m.ptr<uint8_t>());
m.convertTo(m2, CV_8S);
```

Reply:
Firstly, if **roi** is a submatrix, the refcount will add 1. 
`if (this->data == dst.getMat().data && this->elemSize() == CV_ELEM_SIZE(dtype) && refcount == 2) `
 This condition can recognize the first case.
`Mat m (10, 10, CV_8U);
Mat roi = m.rowRange(0, 2);
m.convertTo(roi, CV_8S);`

BUt, I have come up with a new question. Like, 

`Mat m (10, 10, CV_8U);
Mat m2(10, 10, CV_8U, m.ptr<uint8_t>());
m.convertTo(m2, CV_8S);`

In this case,  the member 'u' is NULL. By checking whether 'u' is null, we can solve this problem.

Reply:
`Mat m (10, 10, CV_8U); Mat m2(10, 10, CV_8U, m.ptr<uint8_t>()); m.convertTo(m, CV_8S);`

And, in this case, the original data of 'm' will be released, and 'm2' may encounter unpredictable errors. I believe such risky behavior should be noted by OpenCV users.

Reply:
I have test this case, m2 still operates normally

`Mat m (10, 10, CV_8U); Mat m2(10, 10, CV_8U, m.ptr<uint8_t>()); m.convertTo(m, CV_8S);`

just like

`void test() {
    int * p = new int(5);
    printf("%d\n", *p);
    delete p;
    printf("%d\n", *p);
}`

I think that we shoule make p == nullptr, and in opencv we also should make u == nullptr;

    void deallocate(UMatData* u) const CV_OVERRIDE
    {
        if(!u)
            return;

        CV_Assert(u->urefcount == 0);
        CV_Assert(u->refcount == 0);
        if( !(u->flags & UMatData::USER_ALLOCATED) )
        {
            fastFree(u->origdata);
            u->origdata = 0;
        }
        delete u;
    }